src/hotspot/share/gc/g1/g1HeapRegionChunk.cpp

```diff
@@ -0,0 +1,97 @@
+
+#include "gc/g1/g1CollectedHeap.inline.hpp"
+#include "gc/g1/g1ConcurrentMarkBitMap.inline.hpp"
+#include "gc/g1/g1HeapRegionChunk.hpp"
+#include "gc/g1/heapRegion.hpp"
+
+G1HeapRegionChunk::G1HeapRegionChunk(HeapRegion* region, uint chunk_idx, uint chunk_size, const G1CMBitMap* const bitmap) :
+  _chunk_size(chunk_size),
+  _region(region),
+  _chunk_idx(chunk_idx),
+  _bitmap(bitmap) {
+
+  HeapWord* top = _region->top();
+  HeapWord* bottom = _region->bottom();
+  _start = MIN2(top, bottom + _chunk_idx * _chunk_size);
+  _limit = MIN2(top, bottom + (_chunk_idx + 1) * _chunk_size);
+  _first_obj_in_chunk = _bitmap->get_next_marked_addr(_start, _limit);
+  _next_obj_in_region = _bitmap->get_next_marked_addr(_limit, top);
+  // There is marked obj in this chunk
+  bool marked_obj_in_this_chunk = _start <= _first_obj_in_chunk && _first_obj_in_chunk < _limit;
+  _include_first_obj_in_region = marked_obj_in_this_chunk
+                                 && _bitmap->get_next_marked_addr(bottom, _limit) >= _start;
+}
+
+bool G1ScanChunksInHeapRegions::claim_chunk(uint chunk_id) {
+  return _chunks.par_set_bit(chunk_id);
+}
+
+void G1ScanChunksInHeapRegions::process_chunk(G1HeapRegionChunkClosure* chunk_closure, uint chunk_id, uint worker_id) {
+  G1CollectedHeap* glh = G1CollectedHeap::heap();
+  G1GCPhaseTimes* p = glh->phase_times();
+  Ticks chunk_prepare_start = Ticks::now();
+  uint region_idx = _evac_failure_regions[chunk_id / _chunks_per_region];
+  HeapRegion* r = glh->region_at(region_idx);
+  G1HeapRegionChunk chunk(r, chunk_id % _chunks_per_region, _chunk_size, _bitmap);
+  p->record_or_add_time_secs(G1GCPhaseTimes::PrepareChunks, worker_id, (Ticks::now() - chunk_prepare_start).seconds());
+
+  if (chunk.empty()) {
+    p->record_or_add_thread_work_item(G1GCPhaseTimes::RemoveSelfForwardsInChunks, worker_id, 1, G1GCPhaseTimes::RemoveSelfForwardEmptyChunksNum);
+    return;
+  }
+  p->record_or_add_thread_work_item(G1GCPhaseTimes::RemoveSelfForwardsInChunks, worker_id, 1, G1GCPhaseTimes::RemoveSelfForwardChunksNum);
+  Ticks start = Ticks::now();
+  chunk_closure->do_heap_region_chunk(&chunk);
+  p->record_or_add_time_secs(G1GCPhaseTimes::RemoveSelfForwardsInChunks, worker_id, (Ticks::now() - start).seconds());
+}
+
+G1ScanChunksInHeapRegions::G1ScanChunksInHeapRegions() :
+  _bitmap(G1CollectedHeap::heap()->concurrent_mark()->prev_mark_bitmap()),
+  _chunks(mtGC) { }
+
+void G1ScanChunksInHeapRegions::initialize(const uint* evac_failure_regions, uint evac_failure_regions_length, uint num_workers, const char* task_name) {
+  _evac_failure_regions = evac_failure_regions;
+
+  _chunks_per_region = next_power_of_2(num_workers * G1RemoveSelfForwardPtrsThreadLoadFactor / evac_failure_regions_length);
+  _chunk_size = static_cast<uint>(G1HeapRegionSize / _chunks_per_region);
+  log_debug(gc, ergo)("Running %s using %u workers for removing self forwards with %u chunks per region",
+                      task_name, num_workers, _chunk_size);
+
```

Comment:
```suggestion
  log_debug(gc, ergo)("Initializing removing self forwards with %u chunks per region given %u workers", _chunk_size, num_workers);

```
Since we successfully moved this work into the `G1BatchedGangTask` API, I do not think we need the information about the task and #workers any more. The batch API provides this information afair.
This also allows removal of the `task_name` passed through.